doc/developer/design/20230223_stabilize_with_mutually_recursive.md

```diff
@@ -0,0 +1,297 @@
+
+# Explanation
+[explanation]: #explanation
+<!--
+Explain the design as if it were part of Materialize and you were teaching the team about it.
+This can mean:
+
+- Introduce new named concepts.
+- Explain the feature using examples that demonstrate product-level changes.
+- Explain how it builds on the current architecture.
+- Explain how engineers and users should think about this change, and how it influences how everyone uses the product.
+- If needed, talk though errors, backwards-compatibility, or migration strategies.
+- Discuss how this affects maintainability, or whether it introduces concepts that might be hard to change in the future.
+-->
+
+The original WMR design doc[^wmr] and the technical article[^wmr_article] on our website provide a good explanation about the syntax and semantics of WMR.
+From the implementation point of view, adding WMR support requires extensions across every stage of the query lifecycle.
+The original design doc laid out a plan for adding support in each of the following layers.
+
+1. SQL parsing (&check; in MaterializeInc/materialize#16509)
+2. SQL name resolution (&check; in MaterializeInc/materialize#16509)
+3. SQL planning (&check; in MaterializeInc/materialize#16509)
+4. HIR generalization (&check; in MaterializeInc/materialize#16561)
+5. Lowering (&check; in MaterializeInc/materialize#16561)
+6. MIR generalization (&check; in MaterializeInc/materialize#16561)
+7. MIR optimization corrections (focus of this document)
+8. LIR generalization (&check; in MaterializeInc/materialize#16656, MaterializeInc/materialize#17705)
+9. Rendering (&check; in MaterializeInc/materialize#16787)
+
+The outstanding tasks can be summarized as follows:
+
+1. Complete TODOs that fall under the "MIR optimization corrections".
+1. Enumerate and resolve TODOs in the other stages (if there are any).
+1. Design and execute on a testing plan for the feature.
+1. Design and execute on a rollout plan for the feature.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+<!--
+Focus on the implementation of the feature.
+This is the technical part of the design.
+
+- Is it reasonably clear how the feature is implemented?
+- What dependencies does the feature have and introduce?
+- Focus on corner cases.
+- How can we test the feature and protect against regressions?
+-->
+
+Mostly, the actual work revolves around enumerating and addressing unimplemented code blocks where the corresponding part of our compilation pipeline needs to handle WMR fragments.
+
+## `hir_to_mir` lowering
+
+We might have to check again the changes from MaterializeInc/materialize#16561.
+Mostly, I am concerned is what happens in the presence of:
+
+1. Nested LIR blocks.
+2. Different outer contexts when referencing the same recursive symbol.
+
+## `mir` transformations
```

Comment:
@frankmcsherry: TLDR: I expect the gist of the work to be happening here. I have not done the analysis work so far (tracking issue for that is #17914), but I expect that transforms will be either trivial to fix (just mark them as recursion safe) or to require some rethinking of the analysis or rewrite logic if they are somehow dealing with `LocalId` references. 

For the second group most probably we will see a similar pattern applied all transforms when dealing with recursive symbols, but I would rather do some thinking before I can articulate what this pattern is.

Is there anything else that you I should consider here?